state_machine/drivers/pycubedmini/lib/logging.py

```diff
@@ -0,0 +1,187 @@
+        return string[0:length]
+
+
+def get_logfile_name_dir(logfile_count):
+    logfile_name = get_logfile_name(logfile_count)
+    return f"/sd/logs/{logfile_name}"
+
+
+def get_logfile_name(logfile_count):
+    """
+    format the logfile_name based on logfile_count and return
+    """
+
+    logfile_count_string = rjust(str(logfile_count), 3, '0')
+    return f"log{logfile_count_string}.txt"
+
+
+def new_log(logfile_count):
+    """
+    Create a new log file
+    Write a header and return the new logfile_count
+    """
+
+    # After you fill up 1000 files (~ 1 GB), overwrite old files
+    logfile_count += 1
+    logfile_count %= 1000
+    logfile_name = get_logfile_name(logfile_count)
+    logfile_name_dir = get_logfile_name_dir(logfile_count)
+
+    # if logfile_name doesn't exist, create it
+    if logfile_name not in listdir("/sd/logs/"):
+        # create logfile and write a header
+        logfile = open(logfile_name_dir, "x")
+        logfile.write(f"########## Created: {time.monotonic()} ##########\r\n")
+        logfile.close()
+    # if logfile_name exists, overwrite
+    else:
+        # overwrite existing logfile and write a header
+        logfile = open(logfile_name_dir, "w")
+        logfile.write(f"########## Created: {time.monotonic()} ##########\r\n")
+        logfile.close()
+
+    # print a confirmation message that a new logfile was created
+    write_infotxt(logfile_count)
+
+    return logfile_count
+
+
+def log(msg, max_file_size=max_file_size):
+    """
+    Create/open file and write logs
+    max_file_size is a default value, set to 1E8 at the top of the file
+
+    TODO: find a way to identify the first and last packets of a long log
+    (if we haven't already)
+    - write a timestamp before the first packet
+    - write "\r\n" after the last packet
+    """
+
+    # get logfile_count from info.txt
+    logfile_count = read_infotxt()  # if info.txt doesn't exist, write with 0
+    # get current logfile_name and logfile_name_dir
+    logfile_name_dir = get_logfile_name_dir(logfile_count)
+
+    # if the logs directory is missing, create logs directory and first log
+    if "logs" not in listdir("/sd/"):
+        # create logs directory
+        mkdir("/sd/logs/")
+        # create first logfile
+        logfile = open(logfile_name_dir, "x")
+        logfile.write(f"########## Created: {time.monotonic()} ##########\r\n")
+        logfile.close()
+
+    # if file length + message length is greater than 1MB (1E8)
+    if stat(logfile_name_dir)[6] + len(msg) > max_file_size:
+        # create a new logfile
+        logfile_count = new_log(logfile_count)
+        # get the logfile_name_dir again based on new logfile_count
+        logfile_name_dir = get_logfile_name_dir(logfile_count)
+
+    # open the current logfile and write message msg
+    if cubesat.sd:  # if the SD card is initialized
+        with open(logfile_name_dir, "a+") as file:
+            file.write(f"{msg}")
+
+
+def storage_stats():
```

Comment:
I think we could clean this function up like so:
```python
def storage_usage():
```returns the usage of the SD card and mainboard file system (mainboard, sd)```
sd_usage = None
if cubesat.sd:
    sd_stats = statvfs('/sd')
    sd_usage = sd_stats[3] / sd_stats[2]
main_stats = statvfs('/')
return (main_stats[3] / main_stats[2], sd_usage)
```
I think we only lose information by doing the whole `int(100*actual_value)` thing.